### ISSUE TITLE
An error occurs when ‘max_split_size_mb ’and ‘expandable_segments ’ are enabled at the same time.

### ISSUE THREAD

**Issue by NmomoN** [pytorch/pytorch#123548]
### 🐛 Describe the bug

```python
import os
os.environ["PYTORCH_CUDA_ALLOC_CONF"]='max_split_size_mb:500,expandable_segments:True'
import torch

def create_tensor():
    x = torch.zeros([50,800,1200,3], device='cuda:0')  # 550M     free_block1

if __name__ == '__main__':
    create_tensor()
    temp = []
    number = 500  # Adjusted based on the CUDA memory size.
    for _ in range(number):
        x=torch.zeros([10,800,1200,3], device='cuda:0') # 110MB block2
        temp.append(x)

```
`block2` is continuously allocated and allocated memory close to the upper limit triggers oom. The `release_available_cached_blocks` function is called.But using release_block in this function to release the memory is problematic.
https://github.com/pytorch/pytorch/blob/26bf05ccacc0377f0ef40d1d9c792c403267d5d5/c10/cuda/CUDACachingAllocator.cpp#L2470
https://github.com/pytorch/pytorch/blob/26bf05ccacc0377f0ef40d1d9c792c403267d5d5/c10/cuda/CUDACachingAllocator.cpp#L2500
In `release_block`, the `cudaFree` function is used to release the memory. The correct method is to use the `unmap` function.

In addition, I did not find in the documentation that the two environment variables cannot be shared.

### Versions

PyTorch version: main branch

**Comment by NmomoN** [comment]
@zdevito hello，do you have time to answer my questions?

**Comment by vadimkantorov** [comment]
@zdevito I guess some visual explanations (from the memory_viz of the allocator state) for the https://docs.pytorch.org/docs/stable/notes/cuda.html#optimizing-memory-usage-with-pytorch-cuda-alloc-conf allocator configs would be great to explain the effect of `max_split_size_mb` and `roundup_power2_divisions`. Visualizations of allocator state some test controlled workloads under different allocator configs should maybe augment greatly the text description

**Comment by doplxyz** [comment]
### Self-contained reproducer for the reserved-memory growth

Same interaction on ROCm 7.14 with `torch 2.12.0+rocm7.14.0` (git
`13eab2b3ef4e03999247f063bc2dde01c94e9eaa`), on a 16 GiB RX 9070 XT:

```
PYTORCH_ALLOC_CONF=max_split_size_mb:128,expandable_segments:True
```

`reserved` grows by about 138 MiB per cycle — 220, 2,980, 5,760, 8,520, 11,300,
14,060 MiB — while `allocated` stays between 66 and 306 MiB, until the device is full and
the allocator has to release and retry. Either option alone stays under 500 MiB reserved,
as do defaults.

The growth looks like a pathological interaction between the `max_split_size_mb` filter
in `get_free_block`
([L3436-L3439](https://github.com/pytorch/pytorch/blob/13eab2b3ef4e03999247f063bc2dde01c94e9eaa/c10/cuda/CUDACachingAllocator.cpp#L3436-L3439))
and expandable-segment growth and reuse.

**A note for anyone else trying to reproduce it:** a loop that simply allocates blocks
larger than `max_split_size` and frees them all does *not* reproduce it. The merged free
region ends up adjacent to the segment's unmapped tail, where the expandable-segment path
reuses it happily; reserved plateaued at 1,180 MiB for me that way. What is needed is a
small allocation that stays alive *after* the large ones, fencing the merged free region
off from the tail. Then the too-large check rejects it for the next sub-`max_split_size`
request and the segment grows instead. Controls on the same script: freeing the fence
each cycle, or omitting it, or replacing all the fences with a single up-front allocation
of the same total size, all keep reserved under 500 MiB.

The observed threshold is consistent with the merged block size. With A = 91 MiB and
B = 45.5 MiB freed together (136.5 MiB merged), `max_split_size_mb` of 128 and 136 both
blow up, while 137, 138, 140, 160, 192, 256 and 512 are all stable.

On a real workload (a 4x ESRGAN upscale, no artificial fences) the same pattern appears
on its own: at peak, one expandable segment of 12,160 MiB holds 272 MiB of live
allocations and a single 11,215 MiB inactive block, and 403 of the 451 `segment_map`
calls are triggered by requests smaller than `max_split_size` (22.75 / 45.5 / 68.25 / 91 /
113.75 MiB).

The same reserved-memory growth was observed in a second tested stack combination:
`torch 2.13.0+rocm7.2` (HIP 7.2.53211) on the same card reaches the same 16,000 MiB
high-water mark from the same script, going through 56 release-and-retry rounds while
`allocated` stays under 430 MiB. So this is not confined to one ROCm version or to 2.12.

In the tested 7.14/2.12 combination the churn is additionally followed by corruption of a
live allocation; that was not observed in the tested 7.2/2.13 combination. Because the
PyTorch version and container also differ, this comparison does not attribute the
corruption specifically to ROCm 7.14, and I am not claiming that the data-loss part is in
scope for this issue. I have reported it separately at ROCm/ROCm#6603.

<details>
<summary>repro_vmm_canary.py</summary>

```python
#!/usr/bin/env python3
"""A live allocation reads back as zeros when two allocator options are combined.

Here the damage is the first 20 MiB of the 64 MiB canary in 8 of 10 runs and the whole
64 MiB in the other 2, so the extent is not fixed; the script reports it exactly.

The GPU-side workload consists only of allocations plus one initial `torch.full`.
There is no model, no MIOpen operation, no convolution and no custom kernel.
Validation is a synchronous device-to-host copy followed by a CPU-side comparison.

    PYTORCH_ALLOC_CONF=max_split_size_mb:128,expandable_segments:True \
        python repro_vmm_canary.py

The canary is a 64 MiB buffer, initialised once with 0x5A and never written by user
code again; it is only read back for validation.

Each cycle allocates A = 91 MiB and B = 45.5 MiB, then a small guard block, then frees
A and B. The guard is kept alive, so the merged free region (>= max_split_size) is
trapped in the interior of the expandable segment rather than being adjacent to its
unmapped tail. `get_free_block` skips it for the next sub-128 MiB request because it is
too large, and the expandable-segment path cannot reach it either, so the segment is
grown instead. Reserved memory therefore climbs while the live set stays small.

`--guard-mode` selects the control conditions: `keep` (default) is the failing one,
`free` releases the guard each cycle, `none` allocates no guard, and `ballast` allocates
the full planned aggregate guard size once up front instead of placing one live guard
behind each freed pair. That last one separates "a fence sits between the freed region
and the segment tail" from "more bytes are live".
"""
import argparse, hashlib, json, os
import torch

MiB = 2 ** 20


def read_back(t):
    """One synchronous device-to-host copy. Every check looks at those raw bytes.

    Comparing on the device would leave open the possibility that the reduction
    kernel itself is at fault.
    """
    torch.cuda.synchronize()
    c = t.detach().contiguous().cpu()
    torch.cuda.synchronize()
    return c


def sha(c):
    return hashlib.sha256(c.view(torch.uint8).numpy().tobytes()).hexdigest()


def env_info():
    free_b, total_b = torch.cuda.mem_get_info()
    p = torch.cuda.get_device_properties(0)
    return {
        "alloc_conf": os.environ.get("PYTORCH_ALLOC_CONF", ""),
        "allocator_backend": torch.cuda.get_allocator_backend(),
        "torch": torch.__version__,
        "hip_build": torch.version.hip,
        "gpu": p.name,
        "gcn_arch": getattr(p, "gcnArchName", None),
        "total_mib": round(total_b / MiB),
        "free_mib_before_canary": round(free_b / MiB),
        "hsa_override": os.environ.get("HSA_OVERRIDE_GFX_VERSION", ""),
        "hip_launch_blocking": os.environ.get("HIP_LAUNCH_BLOCKING", ""),
    }


def describe_damage(u, ref_byte=0x5A):
    """Exactly which bytes changed. Reporting only "20 MiB differ" would still allow
    holes inside the first 20 MiB plus damage further up, so state both halves."""
    bad = u != ref_byte
    idx = bad.nonzero(as_tuple=False).flatten()
    n = 20 * MiB
    return {
        "bad_bytes": int(bad.sum()),
        "first_bad_offset": int(idx[0]),
        "last_bad_offset": int(idx[-1]),
        "first_20mib_all_zero": bool((u[:n] == 0).all()),
        "remainder_intact": bool((u[n:] == ref_byte).all()),
        "changed_bytes_all_zero": bool((u[bad] == 0).all()),
    }


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--cycles", type=int, default=200)
    ap.add_argument("--canary-mib", type=int, default=64)
    ap.add_argument("--guard-mode", choices=["keep", "free", "none", "ballast"], default="keep")
    ap.add_argument("--empty-cache-at", type=int, default=-1,
                    help="Call torch.cuda.empty_cache() after this cycle and re-check the "
                         "canary immediately. Lets the unmap path run while the device "
                         "still has plenty of free memory, i.e. with no allocation failure.")
    a = ap.parse_args()

    dev = torch.device("cuda:0")
    # 環境情報は canary を確保する前に採る。あとで採ると free が canary の分だけ
    # 減った値になり、「開始時の空き」と読めなくなる。
    info = env_info()
    canary = torch.full((a.canary_mib * MiB,), 0x5A, dtype=torch.uint8, device=dev)
    ref = sha(read_back(canary))

    info.update({"guard_mode": a.guard_mode, "canary_addr": hex(canary.data_ptr()),
                 "canary_mib": a.canary_mib, "canary_sha256": ref})
    print(json.dumps(info))

    guards = []
    if a.guard_mode == "ballast":
        # The full planned aggregate guard size, allocated once up front, where it
        # cannot fence anything off. Note this is not the same live-bytes curve as
        # the guards, which accumulate one per cycle.
        guards.append(torch.empty(2 * MiB * a.cycles, dtype=torch.uint8, device=dev))

    for i in range(a.cycles):
        A = torch.empty(91 * MiB, dtype=torch.uint8, device=dev)
        B = torch.empty(int(45.5 * MiB), dtype=torch.uint8, device=dev)
        g = None
        if a.guard_mode in ("keep", "free"):
            g = torch.empty(2 * MiB, dtype=torch.uint8, device=dev)
        if i == 0:
            print(json.dumps({"cycle0_addrs": {
                "A": hex(A.data_ptr()), "B": hex(B.data_ptr()),
                "guard": hex(g.data_ptr()) if g is not None else None}}))
        if a.guard_mode == "keep":
            guards.append(g)
        del A, B, g
        torch.cuda.synchronize()

        # Statistics first, then the read-back, so that the counters describe the
        # allocations of this cycle rather than the validation copy.
        st = torch.cuda.memory_stats()
        row = {
            "cycle": i,
            "allocated_mib": round(torch.cuda.memory_allocated() / MiB),
            "reserved_mib": round(torch.cuda.memory_reserved() / MiB),
            "max_reserved_mib": round(torch.cuda.max_memory_reserved() / MiB),
            "num_alloc_retries": st.get("num_alloc_retries", 0),
            "num_ooms": st.get("num_ooms", 0),
        }

        c = read_back(canary)
        row["canary_ok"] = sha(c) == ref
        if not row["canary_ok"]:
            row.update(describe_damage(c.view(torch.uint8)))
            # Read it a second time. If both copies agree, the changed read-back is
            # stable across two synchronous copies rather than varying between reads.
            row["repeat_read_identical"] = bool(torch.equal(c, read_back(canary)))
            print(json.dumps(row))
            raise SystemExit("canary read-back changed although user code never wrote to it")

        if i == a.empty_cache_at:
            # No allocation failure has happened at this point; the device still has
            # room. This isolates the unmap/release path from memory exhaustion.
            free_before = torch.cuda.mem_get_info()[0]
            torch.cuda.empty_cache()
            c2 = read_back(canary)
            ec = {"empty_cache_after_cycle": i,
                  "free_mib_before": round(free_before / MiB),
                  "reserved_mib_after": round(torch.cuda.memory_reserved() / MiB),
                  "num_alloc_retries": torch.cuda.memory_stats().get("num_alloc_retries", 0),
                  "canary_ok": sha(c2) == ref}
            if not ec["canary_ok"]:
                ec.update(describe_damage(c2.view(torch.uint8)))
            print(json.dumps(ec))
            if not ec["canary_ok"]:
                raise SystemExit("canary read-back changed after empty_cache(), with no allocation failure")

        if i % 20 == 0:
            print(json.dumps(row))


if __name__ == "__main__":
    main()
```

</details>

**Comment by doplxyz** [comment]
I have opened a draft PR with a fix for the reserved-memory growth described above: pytorch/pytorch#193400.

The short version: `get_free_block()` applies the `max_split_size` oversize rules to blocks inside expandable segments too. Those rules exist because splitting a standalone `cudaMalloc`ed block keeps the whole block pinned; inside an expandable segment, where pages are mapped and unmapped independently, that does not apply. `should_split()` already ignores `max_split_size` when expandable segments are active, so the PR makes `get_free_block()` consistent with it.

It is a draft because I have not yet run the added regression test against a build that includes the patch.
